### PR TITLE
Keep Track of Global Ordering Requirement

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes #.

## Rationale for this change
During optimization we may remove `SortExec` from the physical plan, if it is satisfied. Once this information is lost, other rules may think, that output ordering is accidental (not a requirement), hence will not try to preserve it. By using an auxiliary executor, we keep track of global ordering requirement across different rules lest to not produce invalid plans.

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/global_order_require.rs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    fn required_input_ordering(&self) -> Vec<Option<Vec<PhysicalSortRequirement>>> {
+        vec![Some(self.requirements.clone())]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        mut children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        assert_eq!(children.len(), 1);
+        let child = children.pop().unwrap();
```
Comment (metesynnada) [inline datafusion/core/src/physical_optimizer/global_order_require.rs:165]:
```suggestion
        let child = children.remove(0);
```

**Review by metesynnada — APPROVED** [review]
LGTM!

**Comment by metesynnada** [comment]
Mustafa, can we update the licence information and rat_exclude for the newly added files?

**Comment by ozankabak** [comment]
Please clear out the merge conflicts and get the change count to a manageable number, I will review then.

**Comment by mustafasrepo** [comment]
> Please clear out the merge conflicts and get the change count to a manageable number, I will review then.

Fixed merge conflict

File: datafusion/core/src/physical_optimizer/global_order_require.rs [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    }
+}
+
+/// Helper function that adds an ancillary `SortRequiringExec` to the given plan.
+fn require_global_ordering(
+    plan: Arc<dyn ExecutionPlan>,
+) -> Result<Arc<dyn ExecutionPlan>> {
+    let mut children = plan.children();
+    // Global ordering defines desired ordering in the final result.
+    // Global ordering can only be propa
```
Comment (ozankabak) [inline datafusion/core/src/physical_optimizer/global_order_require.rs:183]: Please finish this comment, seems accidentally cut-off.

**Review by ozankabak — APPROVED** [review]
LGTM, left one minor comment for you to fix. Once we quickly discuss the tests in the new [join_disable_repartition_joins.slt](https://github.com/synnada-ai/arrow-datafusion/pull/165/files#diff-a96a753b746a1beba3b8d86db9fc4cbbee4997544a3c6ed367a83cd2d4db8568) file, this will be good to go.

**Review by ozankabak — APPROVED** [review]
Ready for merge to our fork 🚀

**Pull request merged**